lib/redis_cluster/client.rb

```diff
@@ -49,33 +59,86 @@ def method_missing(method, *args, &block)
 
     private
 
-    def reload_pool_nodes(raise_error = false)
-      return @pool.add_node!(@startup_hosts, [(0..Configuration::HASH_SLOTS)]) unless @startup_hosts.is_a? Array
+    # Adds only a single node to the client pool and sets it result for the
+    # entire space of slots. This is useful when running either a standalone
+    # Redis or a single-node Redis Cluster.
+    def create_single_node_pool
+      host = @startup_hosts
+      if host.is_a?(Array)
+        if host.length > 1
+          raise ArgumentError, "Can only create single node pool for single host"
+        end
 
-      @mutex.synchronize do
-        @startup_hosts.each do |options|
-          begin
-            redis = Node.redis(@pool.global_configs.merge(options))
-            slots_mapping = redis.cluster("slots").group_by{|x| x[2]}
-            @pool.delete_except!(slots_mapping.keys)
-            slots_mapping.each do |host, infos|
-              slots_ranges = infos.map {|x| x[0]..x[1] }
-              @pool.add_node!({host: host[0], port: host[1]}, slots_ranges)
+        # Flatten the configured host so that we can easily add it to the
+        # client pool.
+        host = host.first
+      end
+
+      @pool.add_node!(host, [(0..Configuration::HASH_SLOTS)])
+    end
+
+    def create_multi_node_pool(raise_error)
+      unless @startup_hosts.is_a?(Array)
+        raise ArgumentError, "Can only create multi-node pool for multiple hosts"
+      end
+
+      @startup_hosts.each do |options|
+        begin
+          redis = Node.redis(@pool.global_configs.merge(options))
+          slots_mapping = redis.cluster("slots").group_by{|x| x[2]}
+          @pool.delete_except!(slots_mapping.keys)
+          slots_mapping.each do |host, infos|
+            slots_ranges = infos.map {|x| x[0]..x[1] }
+            @pool.add_node!({host: host[0], port: host[1]}, slots_ranges)
+          end
+        rescue Redis::CommandError => e
+          if e.message =~ /cluster\ support\ disabled$/
+            if !@force_cluster
+              # We're running outside of cluster-mode -- just create a
+              # single-node pool and move on. The exception is if we've been
+              # asked for force Redis Cluster, in which case we assume this is
+              # a configuration problem and maybe raise an error.
+              create_single_node_pool
+              return
+            elsif raise_error
+              raise e
             end
-          rescue Redis::CommandError => e
-            raise e if raise_error && e.message =~ /cluster\ support\ disabled$/
-            raise e if e.message =~ /NOAUTH\ Authentication\ required/
-            next
-          rescue
-            next
           end
-          break
+
+          raise e if e.message =~ /NOAUTH\ Authentication\ required/
+
+          # TODO: log error for visibility
+          next
+        rescue
+          # TODO: log error for visibility
```

Comment:
There are two TODOs here for now, but I'll send another pull request that adds a logging subsystem and removes them.